server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java

```diff
@@ -55,6 +57,58 @@
 
 public class UpdateRequest extends InstanceShardOperationRequest<UpdateRequest>
         implements DocWriteRequest<UpdateRequest>, WriteRequest<UpdateRequest>, ToXContentObject {
+    private static ObjectParser<UpdateRequest, Void> PARSER;
+
+    private static final ParseField SCRIPT_FIELD = new ParseField("script");
+    private static final ParseField SCRIPTED_UPSERT_FIELD = new ParseField("scripted_upsert");
+    private static final ParseField UPSERT_FIELD = new ParseField("upsert");
+    private static final ParseField DOC_FIELD = new ParseField("doc");
+    private static final ParseField DOC_AS_UPSERT_FIELD = new ParseField("doc_as_upsert");
+    private static final ParseField DETECT_NOOP_FIELD = new ParseField("detect_noop");
+    private static final ParseField FIELDS_FIELD = new ParseField("fields");
+    private static final ParseField SOURCE_FIELD = new ParseField("_source");
+
+    static {
+        PARSER = new ObjectParser<>(UpdateRequest.class.getSimpleName());
+        PARSER.declareField((request, script) -> request.script = script,
+            (parser, context) -> Script.parse(parser), SCRIPT_FIELD, ObjectParser.ValueType.OBJECT_OR_STRING);
+        PARSER.declareBoolean(UpdateRequest::scriptedUpsert, SCRIPTED_UPSERT_FIELD);
+
+        PARSER.declareObject((request, builder) -> request.safeUpsertRequest().source(builder),
+            (parser, context) -> {
+                XContentBuilder builder = XContentFactory.contentBuilder(parser.contentType());
+                builder.copyCurrentStructure(parser);
+                return builder;
+            }, UPSERT_FIELD);
+
+        PARSER.declareObject((request, builder) -> request.safeDoc().source(builder),
+            (parser, context) -> {
+                XContentBuilder docBuilder = XContentFactory.contentBuilder(parser.contentType());
+                docBuilder.copyCurrentStructure(parser);
+                return docBuilder;
+            }, DOC_FIELD);
+
+        PARSER.declareBoolean(UpdateRequest::docAsUpsert, DOC_AS_UPSERT_FIELD);
+        PARSER.declareBoolean(UpdateRequest::detectNoop, DETECT_NOOP_FIELD);
+        PARSER.declareField((request, fields) -> {
+                if (fields != null) {
+                    request.fields(fields.toArray(new String[fields.size()]));
+                }
+            },
+            (parser, context) -> {
+                XContentParser.Token token = parser.currentToken();
+                List<Object> fields = null;
+                if (token == XContentParser.Token.START_ARRAY) {
+                    fields = parser.list();
+                } else if (token.isValue()) {
+                    fields = Collections.singletonList(parser.text());
+                }
+                return fields;
+            }, FIELDS_FIELD, ObjectParser.ValueType.STRING_ARRAY);
```

Comment:
I *think* you can use `PARSER.declareStringArray(UpdateRequest::fields, FIELDS_FIELD)`. It doens't reject non-array things but I think that is fine. Another choice would be to simply pitch this. It has been deprecated for a year and a half (since 1764ec56b38) and we are only going to merge this to the master branch anyway so it is going to be released during a major version so we can make breaking changes like dropping a deprecated field.